File: NN/VNNL/VNNL/VNNL/DenseLayer.cpp
#include "DenseLayer.h"

DenseLayer::DenseLayer(size_t nr_inputs, size_t nr_neurons, float weight_regularizer_l1, float weight_regularizer_l2, float bias_regularizer_l1, float bias_regularizer_l2)
{
    this->weights = 0.01 * xt::random::randn<double>({nr_inputs, nr_neurons});
    this->biases = xt::zeros<double>({1, (int)nr_neurons});
    this->weight_regularizer_l1 = weight_regularizer_l1;
    this->weight_regularizer_l2 = weight_regularizer_l2;
    this->bias_regularizer_l1 = bias_regularizer_l1;
    this->bias_regularizer_l2 = bias_regularizer_l2;
}

void DenseLayer::Forward(xt::xarray<double> inputs)
{
    this->inputs = inputs;

    this->outputs = xt::linalg::dot(inputs, this->weights) + this->biases;
}

// Derivated values is the gradient
void DenseLayer::Backward(xt::xarray<double> derivated_values)
{
    xt::xarray<double> inputs_transposed = xt::transpose(this->inputs);
    this->derivated_weights = xt::linalg::dot(inputs_transposed, derivated_values);

    this->derivated_biases = xt::sum(derivated_values, 0);
    //lambda useful for updating the ones array
    auto regularize = [](xt::xarray<double> weights) {
        auto weights_shape = weights.shape();

        xt::xarray<double> dL = xt::arange(weights.size());
        size_t index = 0;
        for (const auto &value : weights)
        {
            if (value < 0.0)
            {
                dL[index] = -1;
            }
            else
            {
                dL[index] = 1;
            }
            index++;
        }
        dL.reshape(weights.shape());
        return dL;
    };
    regularize(weights);
    //L1 on weights
    if (this->weight_regularizer_l1 > 0)
    {

        xt::xarray<double> dL1 = regularize(this->weights);
        this->derivated_weights += this->weight_regularizer_l1 * dL1;
    }
    // L2 on weights
    if (this->weight_regularizer_l2 > 0)
    {
        this->derivated_weights += 2 * this->weight_regularizer_l2 * this->weights;
    }
    // L1 on biases
    if (this->bias_regularizer_l1 > 0)
    {
        xt::xarray<double> dL1 = regularize(this->biases);
        this->derivated_biases += this->bias_regularizer_l1 * dL1;
        cout<<this->derivated_biases;
    }
    // L2 on biases
    if (this->bias_regularizer_l2 > 0)
    {
        this->derivated_biases += 2 * this->bias_regularizer_l2 * this->biases;
    }
    //   Gradient on values
    xt::xarray<double> weights_transposed = xt::transpose(this->weights);
    this->derivated_inputs = xt::linalg::dot(derivated_values, weights_transposed);
}

std::pair<xt::xarray<double>, xt::xarray<double>> DenseLayer::getParameters()
{
    return std::pair<xt::xarray<double>, xt::xarray<double>>(this->weights, this->biases);
}
void DenseLayer::setParameters(xt::xarray<double> weights, xt::xarray<double> biases)
{
    this->weights = weights;
    this->biases = biases;
}

File: Music Recognizer/Layers/DenseLayer.py
import numpy as np


class DenseLayer:
    """
        DenseLayer/FullyConectedLayer adds a non-linearity propriety to the network and performs the
        learning by multiplying the inputs to the weights, then adding the biases.

    The DenseLayer class contains:
        :param weights(np.array) : Dense Layer's weights.
        :param biases(np.array) : Dense Layer's biases
        :param output(np.array) : The outputs of the layer after the forward pass
        :param weight_regularizer_l1(float) : l1 weight regulizer factor
        :param weight_regularizer_l2(float) : l2 weight regulizer factor
        :param bias_regularizer_l1(float) : l1 bias regulizer factor
        :param bias_regularizer_l2(float) : l2 bias regulizer factor
        :param inputs(np.array) : inputs array
        :param derivated_inputs(np.array) : The outputs of the layer after the backward pass

        Sources:    * Neural Networks from Scratch - Harrison Kinsley & Daniel Kukieła [pg.66-71]

    """
    def __init__(self, inputs_number, neuron_number,
                 weight_regularizer_l1=0, weight_regularizer_l2=0,
                 bias_regularizer_l1=0, bias_regularizer_l2=0):
        self.weights = 0.01 * np.random.randn(inputs_number, neuron_number)
        self.biases = np.zeros((1, neuron_number))
        self.inputs = None
        self.output = None
        self.weight_regularizer_l1 = weight_regularizer_l1
        self.weight_regularizer_l2 = weight_regularizer_l2
        self.bias_regularizer_l1 = bias_regularizer_l1
        self.bias_regularizer_l2 = bias_regularizer_l2

    def forward(self, inputs):
        """
        Performs the forward pass by multipyling the weights with the inputs then adding the biases.
        Args:
            inputs (np.array): given inputs.

        Sources:    * Neural Networks from Scratch - Harrison Kinsley & Daniel Kukieła [pg.66-71]
        """
        # Storing the input values
        self.inputs = inputs
        # Apply the formula
        self.output = np.dot(inputs, self.weights) + self.biases

    def backward(self, derivated_values):
        """
        Performs the backward pass by using the gradient chaining method.

        Args:
            derivated_values (np.array): Derivated inputs .

        Sources:    * Neural Networks from Scratch - Harrison Kinsley & Daniel Kukieła [pg.66-71]
        
        """
        # Computing the derivated values
        self.derivated_weights = np.dot(self.inputs.T, derivated_values)
        self.derivated_biases = np.sum(derivated_values, axis=0, keepdims=True)
        # Applying gradients in regards to regularisation if they re initialized
        if self.weight_regularizer_l1 > 0:
            dL1 = np.ones_like(self.weights)
            dL1[self.weights < 0] = -1
            self.derivated_weights += self.weight_regularizer_l1 * dL1
        if self.weight_regularizer_l2 > 0:
            self.derivated_weights += 2 * self.weight_regularizer_l2 * \
                             self.weights
        if self.bias_regularizer_l1 > 0:
            dL1 = np.ones_like(self.biases)
            dL1[self.biases < 0] = -1
            self.derivated_biases += self.bias_regularizer_l1 * dL1
        if self.bias_regularizer_l2 > 0:
            self.derivated_biases += 2 * self.bias_regularizer_l2 * \
                            self.biases
        # Computing the output
        self.derivated_inputs = np.dot(derivated_values, self.weights.T)

    def get_parameters(self):
        """
        Return the layer's current parameters. 
        
        Sources:    * Neural Networks from Scratch - Harrison Kinsley & Daniel Kukieła [pg.66-71]
        """
        return self.weights, self.biases